EVP & TLS: Add necessary EC_KEY data extraction functions, and use them

levitte opened the pull request:
libssl code uses EVP_PKEY_get0_EC_KEY() to extract certain basic data
from the EC_KEY.  We replace that with EVP_PKEY functions.

This may or may not be refactored later on.

levitte committed:
EVP: add EVP_PKEY_is_a() and EVP_PKEY_can_sign()

EVP_PKEY_is_a() is the provider side key checking function corresponding
to checking EVP_PKEY_id() or an EVP_PKEY against macros like EVP_PKEY_EC.
It also works with legacy internal keys.

We also add a warning indoc/man3/EVP_PKEY_set1_RSA.pod regarding the
reliability of certain functions that only understand legacy keys.

Finally, we take the opportunity to clean up doc/man3/EVP_PKEY_set1_RSA.pod
to better conform with man-page layout norms, see man-pages(7) on Linux.

levitte committed:
EVP & TLS: Add necessary EC_KEY data extraction functions, and use them

libssl code uses EVP_PKEY_get0_EC_KEY() to extract certain basic data
from the EC_KEY.  We replace that with internal EVP_PKEY functions.

This may or may not be refactored later on.

levitte commented:
If nothing else, this can be an interim solution until #11340 has been merged.
#11328 depends on something like this.

mattcaswell commented:
Hmmm. I can't say I'm keen on this. This seems to just sidestep the problem of what to do with the `EVP_PKEY_get0_EC_KEY` et al functions - and fixes it *for us*. But does so by polluting the EVP_PKEY API with algorithm specific functions.

levitte commented:
Yes, that's true.  However, as soon as we get EC key generation, we get issues with this, so the question is how long that should be stuck waiting for "the right thing" to happen.

I *could* make those functions internal but still exported, if that helps matters

levitte commented:
This *is* an interim solution

levitte commented:
Marked as WIP for obvious reasons

mattcaswell commented:
> I could make those functions internal but still exported, if that helps matters

That would definitely help I think.

levitte commented:
Although, I think `EVP_PKEY_can_sign()` is more generally useful...

levitte commented:
There.  I've also added EVP_PKEY_is_a(), which has been floating a more than one branch I juggled with

mattcaswell commented on doc/man3/EVP_PKEY_is_a.pod:
```diff
@@ -0,0 +1,42 @@
+
+=head1 DESCRIPTION
+
+EVP_PKEY_is_a() checks if the key type of I<pkey> is I<name>.
```
I think some examples of valid names would be useful.

levitte replied:
Sure

levitte replied:
Done

mattcaswell commented on ssl/statem/statem_lib.c:
```diff
@@ -1559,10 +1558,9 @@ static int is_tls13_capable(const SSL *s)
-        curve = EC_GROUP_get_curve_name(EC_KEY_get0_group(eckey));
+        curve = EVP_PKEY_get_EC_KEY_curve_nid(s->cert->pkeys[SSL_PKEY_ECC]
+                                              .privatekey);
+        fprintf(stderr, "DEBUG[%s] ECC curve = %d\n", __func__, curve);
```
Err...fprintf?

levitte replied:
Oops, debug stuff.  Will remove

levitte replied:
Removed them all

mattcaswell commented on ssl/statem/statem_lib.c:
```diff
@@ -1559,10 +1558,9 @@ static int is_tls13_capable(const SSL *s)
-        if (eckey == NULL)
-            continue;
-        curve = EC_GROUP_get_curve_name(EC_KEY_get0_group(eckey));
+        curve = EVP_PKEY_get_EC_KEY_curve_nid(s->cert->pkeys[SSL_PKEY_ECC]
```
Shouldn't this be `evp_pkey_get_EC_KEY_curve_nid` (i.e. in lowercase)?

levitte replied:
Yes.  There are a few of these, I've already corrected them locally

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -577,19 +577,15 @@ void tls1_get_formatlist(SSL *s, const unsigned char **pformats,
 
     /* Get required compression id */
-    if (EC_KEY_get_conv_form(ec) == POINT_CONVERSION_UNCOMPRESSED) {
+    if (EVP_PKEY_get_EC_KEY_conv_form(pkey) == POINT_CONVERSION_UNCOMPRESSED) {
```
evp_pkey_...and so on for other occurrences

mattcaswell commented on include/internal/evp.h:
```diff
@@ -0,0 +1,18 @@
+
+# include <openssl/evp.h>
+
+int evp_pkey_get_EC_KEY_curve_nid(const EVP_PKEY *pkey);
```
Should there be some TODO comment somewhere saying this is an interim solution?

levitte replied:
I thought I already added such a comment???  probably somewhere else, and yeah I can duplicate it here

levitte replied:
Done

mattcaswell commented on include/internal/evp.h:
```diff
@@ -0,0 +1,18 @@
+ */
+
+#ifndef OSSL_INTERNAL_DSO_H
+# define OSSL_INTERNAL_DSO_H
```
Should be `OSSL_INTERNAL_EVP_H`

levitte replied:
Yup, already fixed locally

mattcaswell commented:
Could you push the latest version of your local branch here? I'm experimenting with your ec key gen stuff...

levitte commented:
Done, and in #11328 as well

levitte committed:
Fix export of provided EC keys

The exporter freed a buffer too soon, and there were attempts to use
its data later, which was overwritten by something else at that
point.

levitte committed:
TLS: Temporarly downgrade newly generated EVP_PKEYs to legacy

The transfer of TLS encodedpoint to backends isn't yet fully supported
in provider implementations.  This is a temporary measure so as not to
get stuck in other development.

levitte commented:
Rebased on master, now that #11375 is in.

levitte commented:
Ping!

levitte commented:
Travis failure re documentation:

```
doc/man3/EVP_PKEY_is_a.pod:1: EXAMPLE not EXAMPLES section.
```

That document contains *two* examples, so I'm starting to wonder if find-doc-nits isn't a little *too* picky in this case.

levitte commented:
Ping!

mattcaswell commented:
Needs rebase!

t8m commented:
There are now conflicts and the Travis failure is probably relevant

levitte commented:
Just rebased

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -577,19 +578,15 @@ void tls1_get_formatlist(SSL *s, const unsigned char **pformats,
 /* Check a key is compatible with compression extension */
 static int tls1_check_pkey_comp(SSL *s, EVP_PKEY *pkey)
```
This function is inextricably linked with the point encoding/decoding. When we push the point encoding/decoding provider side then we will also have to push the logic about what point compression is supported provider side too, i.e. we should ask the provider what point compression is supported for any particular key - rather than trying to infer if from the field type. That works for own stuff because we know we support point compression - but we can't necessarily assume the same for provider owned keys. For now though we are explicitly downgrading EC keys prior to point encoding/decoding in order to use the ameth implementation. It seems we could do the same trick here and then just do it the old way - until such time as the whole thing goes provider side.

levitte replied:
Okie, done

mattcaswell commented on doc/man3/EVP_PKEY_is_a.pod:
```diff
@@ -0,0 +1,74 @@
+
+EVP_PKEY_is_a() checks if the key type of I<pkey> is I<name>.
+
+EVP_PKEY_can_sign() checks if I<pkey> can be used for signing.  The
```
I assume this returns false if the key can only be used for verifying? e.g. because it is a public key? Could we make that more explicit?

levitte replied:
Ah, so having it verify that `evp_keymgmt_has(p->keymgmt, p->keydata, OSSL_KEYMGMT_SELECT_PRIVATE_KEY)` is true?  Yeahsure

mattcaswell replied:
See my comment above...

mattcaswell commented on doc/man3/EVP_PKEY_set1_RSA.pod:
```diff
@@ -51,52 +51,61 @@ EVP_PKEY_set1_engine, EVP_PKEY_get0_engine - EVP_PKEY assignment functions
+The following functions are only reliable with B<EVP_PKEY>s that have
+been assigned an internal key with EVP_PKEY_assign_*():
+
+EVP_PKEY_id(), EVP_PKEY_base_id(), EVP_PKEY_type(), EVP_PKEY_set_alias_type()
```
Should we attempt a downgrade if someone calls these functions?

levitte replied:
I'm unsure...  but as a side thought, I've been thinking that keys with a provider side "origin" should have the type `EVP_PKEY_KEYMGMT` rather than `EVP_PKEY_NONE`...

levitte replied:
I'm leaning on no.  We risk too much uncontrolled downgrading.

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -732,6 +736,110 @@ int EVP_PKEY_base_id(const EVP_PKEY *pkey)
+            : evp_first_name(prov, pkey->keymgmt->name_id);
+        signature = EVP_SIGNATURE_fetch(libctx, supported_sig, NULL);
+        EVP_SIGNATURE_free(signature);
+        return signature != NULL;
```
A pointer is just a value just like any other value.  That value won't "magically" change because you freed the memory area its value points at.  The value in itself is perfectly usable as long as it's not dereferenced.

But, I can waste another variable to hold the truth value if you prefer...

richsalz replied:
FWIW, I agree that looking at the pointer after it's free'd, while technically legal, is confusing.

kaduk replied:
I think that *technically* it's not actually legal.
It helps to recall that the C rules for pointers have their origins in the era of segmented architectures, where even trying to load a pointer value (not dereference it) for which the corresponding segment was not valid anymore, could fault.

levitte replied:
Huh!  I didn't know that.

t8m replied:
You could nicely do 
```
if (signature != NULL) {
    EVP_SIGNATURE_free(signature); 
    return 1; 
} return 0;
```

richsalz replied:
I believe Ben is wrong; "free(p); look at p but not *p" is legal.

kaduk replied:
See n1256.pdf appendix J.2, stanzas 8 and 9:

```
— An object is referred to outside of its lifetime (6.2.4).
— The value of a pointer to an object whose lifetime has ended is used (6.2.4).
```
The first one refers to dereferencing the pointer, and the second refers to only the value of the (undereferenced) pointer.

richsalz replied:
I stand corrected, thanks.

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -732,6 +736,110 @@ int EVP_PKEY_base_id(const EVP_PKEY *pkey)
+            pkey->keymgmt->query_operation_name != NULL
+            ? pkey->keymgmt->query_operation_name(OSSL_OP_SIGNATURE)
+            : evp_first_name(prov, pkey->keymgmt->name_id);
+        signature = EVP_SIGNATURE_fetch(libctx, supported_sig, NULL);
```
Looking at the legacy code, it seems we assume that method that supports signing also supports verifying.  Looking at what `EV_KEY_can_sign()` does, I found these indicators:

https://github.com/openssl/openssl/blob/20c98cd45399423f760dbd75d8912769c6b7b10e/crypto/ec/ec_local.h#L31-L32

https://github.com/openssl/openssl/blob/20c98cd45399423f760dbd75d8912769c6b7b10e/include/openssl/evp.h#L1402-L1404

levitte replied:
Should we care in the legacy section of this function?  How do we check if there's a private key there?

levitte replied:
The question here remains unanswered.  @mattcaswell

mattcaswell replied:
It seems all this stems from confusion in my own head about what EVP_PKEY_can_sign() is supposed to do. Is it supposed to report whether the parameters associated with this key are capable of doing signing operations, or whether this specific key can do signing (i.e. it has a private key *and* it has suitable parameters). I was assuming the latter - but if this is supposed to be a drop-in replacement for EC_KEY_can_sign() then that seems to do the former - and therefore don't care whether this has a private key or not.

levitte replied:
I've designed it as a drop in replacement for `EC_KEY_can_sign()`.  It's possible that we want to make it stricter before releasing 3.0, though...  but maybe not immediately, as we seem to get ourselves into another philosophical rabbit hole.

levitte replied:
Fixed

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -732,6 +736,110 @@ int EVP_PKEY_base_id(const EVP_PKEY *pkey)
+        case EVP_PKEY_ED25519:
+        case EVP_PKEY_ED448:
+        case EVP_PKEY_EC:        /* Including SM2 */
+            return EC_KEY_can_sign(pkey->pkey.ec);
```
Okie

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -732,6 +736,110 @@ int EVP_PKEY_base_id(const EVP_PKEY *pkey)
+}
+
+#ifndef OPENSSL_NO_EC
+/* TODO rewrite when we have proper data extraction functions */
```
I wasn't ready to assume, so...

(I assume, BTW, that we're going to look at *all* TODO comments before final release, and that the presence of "(3.0)" only indicates if it's a "MUST do" or a "MAY do")

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -577,19 +578,26 @@ void tls1_get_formatlist(SSL *s, const unsigned char **pformats,
+     */
+    EVP_PKEY_get0(pkey);
+    if (EVP_PKEY_id(pkey) == EVP_PKEY_NONE)
+        return 0;
```
Now that we have downgraded we can keep all the original code below. The `EVP_PKEY_get0_EC_KEY()` call will just work and we don't need `evp_pkey_get_EC_KEY_conv_form` or `evp_pkey_get_EC_KEY_field_type`.

levitte replied:
At *some* point, we will need to get those things from provider side keys, so I'm apprehensive at removing those functions...  do you anticipate that the *final solution* will look substantially different?

levitte replied:
Ping @mattcaswell

mattcaswell replied:
Yes. This whole dance with these calls is just to figure out what TLS compression id we need to use. I anticipate asking the provider directly for this compression id rather than inferring it from these other calls.

kaduk replied:
(I assume "TLS compression id" means "EC point format", not "we're using something other than NULL compression".)

mattcaswell replied:
Yes...the TLS EC point format. Sorry for the confusion.

levitte replied:
In that case, I might as well restore this whole function to its original form, and just replace the EVP_PKEY_id() check with a EVP_PKEY_is_a() check.

levitte replied:
Done

levitte commented:
Ah, the CI failure was due to a merge error with duplicate definition of the same local variable.
Fixed, and I took the opportunity to squash and rebase.

mattcaswell commented:
Approved assuming CIs pass.

levitte commented:
... it only takes a rebase for hell to break loose :unamused:

levitte committed:
fixup! EVP: add EVP_PKEY_is_a() and EVP_PKEY_can_sign()

levitte commented:
With @mattcaswell's assistance, I was able to put in a workable hack.  It turned out that legacy key support needed some hard coding.

levitte commented:
One failure left...  I didn't account for `no-ec`

levitte commented:
... at least not everywhere

levitte committed:
fixup! EVP: add EVP_PKEY_is_a() and EVP_PKEY_can_sign()

levitte committed:
fixup! EVP & TLS: Add necessary EC_KEY data extraction functions, and use them

levitte commented:
Fixed

mattcaswell reviewed (approved):
Reconfirm assuming CIs are ok

levitte commented:
Apart from the usual timeout, CIs are fine

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

levitte commented:
Merged.

4f76d62f2384d3335bd1d043706995ae64b37348 EVP: add EVP_PKEY_is_a() and EVP_PKEY_can_sign()
c2041da8c15027ddde5afcf9809d8d3a975eb25b EVP & TLS: Add necessary EC_KEY data extraction functions, and use them
e3be0f4389fe1680e1fd547a8575ea71b4518b57 Fix export of provided EC keys
afce590b74159f7df1452fb2c4aa990a52536c38 TLS: Temporarly downgrade newly generated EVP_PKEYs to legacy

levitte commented:
Oh damn, *now* I see that the last approval came after the "approval: done" label was set, so I cut the grace period short...  by 3hrs, I hope it's not too bad.
